### DISCUSSION TITLE
Retry download from dist URL

### DISCUSSION THREAD

**Discussion by adamwilson919** [composer/composer#11411] [Q&A / Support] [answered]
Hi all,

Is it possible to retry a package download from the dist URL (v2.0.0)? At the moment, if it fails it reverts to the source URL.

- https://github.com/composer/composer/blob/2.0.0/src/Composer/Downloader/DownloadManager.php#L194

We're using our own artifact repo for the dist source and sometimes (very rarely!), composer gets a 404 from the dist URL during our build pipeline. It seems temp network related, as other packages in the same install download fine. Re-running composer fixes it. Our pipeline doesn't have external access, so the source is pointless.

Is there a way to ignore the source URL completely and retry the dist URL? Alternatively, is there a command to validate all installed packages are downloaded?

The added complication is that we didn't get an exit code when composer failed to download from dist & source.

**Comment by stof** [comment]
> The added complication is that we didn't get an exit code when composer failed to download from dist & source.

This is weird, as composer should be failing if neither dist nor source could be downloaded

**Reply by adamwilson919** [reply]
Yeah I thought the same, example output below. For package 76 it fails, then tries from source but fails as there's no outbound network connectivity (and the package doesn't exist when double checking vendors). Yet it doesn't report the failure.

I'll add verbose logging and monitor. Although it's one of them 'fun' ones which only happens intermittently.

```
$ /usr/local/bin/composer install -v -d .
Installing dependencies from lock file (including require-dev)
Verifying lock file contents can be installed on current platform.
Dependency resolution completed in 0.000 seconds
Package operations: 79 installs, 0 updates, 0 removals
Installs: behat/gherkin:v4.6.2, ...[omitted]..., symfony/property-access:v3.0.9, symfony/serializer:v3.0.9, tedivm/stash:v0.14.1
  - Downloading behat/gherkin (v4.6.2)
  - Downloading ...[omitted]
  - Downloading symfony/property-access (v3.0.9)
  - Downloading symfony/serializer (v3.0.9)
  - Downloading tedivm/stash (v0.14.1)
  0/79 [>---------------------------]   0% < 1 sec
  1/79 ...[omitted]...
 74/79 [==========================>-]  93% 49 secs
 76/79 [==========================>-]  96% 54 secs
 Failed to download symfony/property-access from dist: The "https://[local-url].com/vcs-dists/zip/symfony/property-access/a4f1a668d0a269a6579" file could not be downloaded (HTTP/1.1 404 Not Found)
    Now trying to download from source
 77/79 [===========================>]  97%  1 min
 78/79 [===========================>]  98%  1 min
 79/79 [============================] 100%  1 min
  - Installing behat/gherkin (v4.6.2): Extracting archive
  - Installing ...[omitted]...
  - Installing symfony/serializer (v3.0.9): Extracting archive
  - Installing tedivm/stash (v0.14.1): Extracting archive
  0/63 [>---------------------------]   0% < 1 sec
  8/63 [===>------------------------]  12% < 1 sec
 17/63 [=======>--------------------]  26% < 1 sec
 27/63 [============>---------------]  42% < 1 sec
 36/63 [================>-----------]  57% < 1 sec
 45/63 [====================>-------]  71%  1 sec
 55/63 [========================>---]  87%  1 sec
 59/63 [==========================>-]  93%  1 sec
 63/63 [============================] 100%  1 sec
Generating autoload files
```

**Reply by adamwilson919** [reply, marked as answer]
I think I've found the cause. We don't have git installed on the Docker image we use for builds. Composer v2.0.0 (the one we're using) doesn't throw an exception when git isn't available. So silently fails on git source downloads.

- https://github.com/composer/composer/blob/2.0.0/src/Composer/Downloader/GitDownloader.php#L59

The latest code does though:

- https://github.com/composer/composer/blob/main/src/Composer/Downloader/GitDownloader.php#L82

I'll arrange for composer to be upgraded and hopefully that'll resolve it!
